dali/pipeline/operators/color/hsv.h

```diff
@@ -0,0 +1,217 @@
+extern const std::string kValue;
+extern const std::string kOutputType;
+
+/**
+ * Color space conversion
+ */
+const mat3 Rgb2Yiq = {{
+                              {.299f, .587f, .114f},
+                              {.596f, -.274f, -.321f},
+                              {.211f, -.523f, .311f}
+                      }};
+
+// Inversion of Rgb2Yiq, but pre-calculated, cause mat<> doesn't do inversion.
+const mat3 Yiq2Rgb = {{
+                              {1, .956f, .621f},
+                              {1, -.272f, -.647f},
+                              {1, -1.107f, 1.705f}
+                      }};
+
+
+inline mat3 compose_hue(float hue /* hue hue hue */ ) {
+  const auto h_rad = hue * M_PI / 180;
+  mat3 ret = mat3::eye();  // rotation matrix
+  // Hue change in YIQ color space is a rotation along the Y axis
+  ret(1, 1) = cos(h_rad);
+  ret(2, 2) = cos(h_rad);
+  ret(1, 2) = -sin(h_rad);
+  ret(2, 1) = sin(h_rad);
+  return ret;
+}
+
+inline mat3 compose_saturation(float saturation) {
+  mat3 ret = mat3::eye();
+  // In the YIQ color space, saturation change is a
+  // uniform scaling in IQ dimensions
+  ret(1, 1) = saturation;
+  ret(2, 2) = saturation;
+  return ret;
+}
+
+
+inline mat3 compose_value(float value) {
+  // In the YIQ color space, value change is a
+  // uniform scaling across all dimensions
+  return mat3::diag(value);
+}
+
+}  // namespace hsv
+
+
+
+template <typename Backend>
+class Hsv : public Operator<Backend> {
+ public:
+  ~Hsv() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(Hsv);
+
+ protected:
+  explicit Hsv(const OpSpec &spec) :
+          Operator<Backend>(spec),
+          output_type_(spec.GetArgument<DALIDataType>(hsv::kOutputType)) {
+    GetSingleOrRepeatedArg(spec, hue_, hsv::kHue, batch_size_);
+    GetSingleOrRepeatedArg(spec, saturation_, hsv::kSaturation, batch_size_);
+    GetSingleOrRepeatedArg(spec, value_, hsv::kValue, batch_size_);
+    tmatrices_ = determine_transformation(hue_, saturation_, value_);
+    if (std::is_same<Backend, GPUBackend>::value) {
+      kernel_manager_.Resize(1, 1);
+    } else {
+      kernel_manager_.Resize(num_threads_, batch_size_);
+    }
+    assert(hue_.size() == static_cast<size_t>(batch_size_));
+    assert(saturation_.size() == static_cast<size_t>(batch_size_));
+    assert(value_.size() == static_cast<size_t>(batch_size_));
+  }
+
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+
+  /**
+   * @brief Creates transformation matrices based on given args
+   */
+  std::vector<mat3>
+  determine_transformation(const std::vector<float> &hue, const std::vector<float> &saturation,
+                           const std::vector<float> &value) const {
+    using namespace hsv;  // NOLINT
+    std::vector<mat3> ret;
+
+    auto size = hue.size();
+    ret.resize(size);
+    for (size_t i = 0; i < size; i++) {
+      ret[i] = Yiq2Rgb * compose_hue(hue[i]) * compose_saturation(saturation[i]) *
+               compose_value(value[i]) * Rgb2Yiq;
```

Comment:
Why do you insist on having these matrix multiplications here? Especially the perfectly avoidable compose_value (you'd be better off just multiplying by the scalar). And, as I demonstrated before, the function that does all three transforms is not that complicated, either.

Reply:
I think that this is ok from readability standpoint. You define those transformations as rotations/scaling/shifting and compose them into one transformation.

Reply:
When we're at readability: how about:
```
mat3 color_transform = rotate_hue(hue[i]) * scale_saturation(saturation[i]) * value[i];
return Yiq2Rgb * color_transform * Rgb2Yiq;
```
These `compose_xxx` names don't really tell us what we do with `xxx`.

Reply:
Well, `rotate_hue` isn't the best name either, cause it doesn't actually rotate anything.
How about `compose_hue_matrix`? Or sth shorter:
```
/**
 * Composes transformation matrix for hue
 */
mat3 hue_mat(){}
[...]

return Yiq2Rgb * hue_mat(hue[i]) * sat_mat(saturation[i]) * val_mat(value[i]) * Rgb2Yiq;
```
?